pkg/virt-handler/vm.go

```diff
@@ -126,44 +126,70 @@ func (d *VMHandlerDispatch) getVMNodeAddress(vm *v1.VirtualMachine) (string, err
 	return addrStr, nil
 }
 
-func (d *VMHandlerDispatch) updateVMStatus(vm *v1.VirtualMachine, cfg *api.DomainSpec) error {
-	obj, err := scheme.Scheme.Copy(vm)
-	if err != nil {
-		return err
+func (d *VMHandlerDispatch) updateVMStatus(vm *v1.VirtualMachine, isPending bool, newDomainSpec *api.DomainSpec, syncError error) error {
+
+	var err error
+	oldPhase := vm.Status.Phase
+
+	// TODO, update a VM condition, right now, do nothing
+	if syncError != nil {
+		return nil
+	}
+
+	// Don't update the VM if it is already in a final state
+	if vm.IsFinal() {
+		return nil
 	}
-	vm = obj.(*v1.VirtualMachine)
 
-	// XXX When we start supporting hotplug, this needs to be altered.
-	// Check if the VM is already marked as running. If yes, don't update the VM.
-	// Otherwise we end up in endless controller requeues.
-	if vm.Status.Phase == v1.Running {
+	// While the VM is migrating, don't do anything, the Migration Controller is in charge
+	if vm.Status.MigrationNodeName != "" {
+		return nil
+	}
+
+	// Failed/Succeeded is currently handled somewhere else, do nothing if the spec is not there
+	if newDomainSpec == nil {
 		return nil
 	}
 
-	vm.Status.Phase = v1.Running
+	// TODO update to Failed if VM e.g. crashes
+	// TODO update based on api.Domain succeeded/failed here, instead of the extra controller
+	if isPending {
```

Comment:
It's likely this will cause some unexpected behavior in the future.

I didn't intend this boolean to mean anything with regards to the VM's phase. For example, I'm using it in #526 during graceful shutdown to indicate the VM should be processed again soon. It's possible to set this boolean during any async action throughout the lifecycle of the vm.

Reply:
#530 provides the whole story (follow up PR). isPending is not needed there anymore. I think that we should directly enqueue on the processVmStatus, and not return isPending at all, like I did there. I will move that part over.

Reply:
Actually, I can't, until we unify the controller in #530. The isPending still does here what it does right now in the code, it delays the status update (which is per se not right, I agree).

Reply:
with the logic you have in #530, we don't need the isPending boolean to be passed to the update status function anymore.

@rmohr what's your plan here? Are you going to abandon this patch in favor of #530?  I really like #530

Reply:
Since #530 is built based on this PR, I would prefer, if we merge this one first, since it fixes a few things immediately and it reduces the size of #530. I will make #530 mergabel asap, afterwards.

Reply:
can we just roll it all into one pr?